### ISSUE TITLE
Running dotnet core on Alpine for ARM32 

### ISSUE THREAD

**Issue by JensNordenbro** [dotnet/runtime#2021]
Due to its small footprint (OS image size), Alpine is suited for embedded scenarios on ARM32.

Things leading to this:

+ Today dotnet core is supported on ARM32.
+ Today dotnet core 2.1 is running on Alpine for x86.
+ Today dotnet core 3.1 is running on Alpine for ARM64.

It is expected to have symmetry in OS support between different CPU architectures where possible.

(If/when this is implemented the docker images for running .net core application should be available. )

This is a fork of https://github.com/dotnet/coreclr/issues/17468

**Comment by ericstj** [comment]
Current support matrix https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1-supported-os.md#linux

It seems this is a request to add Alpine ARM32 to this.  @leecow / @jeffschwMSFT do you know if this is something that's been considered?  Are there any technical blockers?  /cc @wfurt 

Marking as 5.0 to answer the question.

**Comment by janvorli** [comment]
@ericstj There are no technical blockers. It is just about wiring it to our build lab / docker containers (whcih includes rootfs creation). Maybe RID graph, I'm not sure whether we already have it covered or not. Otherwise I believe it should just work.

**Comment by JensNordenbro** [comment]
Alsö @ericstj ; work was done on 17468 and we were a couple of people following the issue and still wanted!

**Comment by ericstj** [comment]
Alpine is only x64 in rid graph: https://github.com/dotnet/runtime/blob/129ee5252757ec8bdfccf9b271ded36675570309/src/libraries/pkg/Microsoft.NETCore.Platforms/runtimeGroups.props#L16-L20

Seems like that's not even consistent with the matrix.

**Comment by janvorli** [comment]
That's strange as we've supported Alpine on ARM64 since 3.0 (see https://hub.docker.com/_/microsoft-dotnet-core-runtime-deps/, the Linux arm64 tags)

**Comment by ericstj** [comment]
> Alsö @ericstj ; work was done on 17468 in the old repo.

Looks to be https://github.com/dotnet/runtime/issues/10111 in new repo.

**Comment by ericstj** [comment]
> That's strange as we've supported Alpine on ARM64 since 3.0 (see https://hub.docker.com/_/microsoft-dotnet-core-runtime-deps/, the Linux arm64 tags)

Looking at the product assets I see it was treated as `linux-musl-arm64` so that works from shared-fx but would fail for folks trying to resolve from `alpine.ver-arm64` in a self-contained publish.

**Comment by hamish-milne** [comment]
Is it possible to get a build of this working at present, perhaps using the instructions here https://github.com/dotnet/runtime/issues/10111#issuecomment-384370445 ? I've tried publishing a simple app with the `linux-musl-arm` RID, which works, but execution fails with `Illegal instruction` at `__stack_chk_fail` (some incompatibility with alpine's stack-protection-by-default?)

Happy to help out with this as it's very relevant to my interests :)

**Comment by hamish-milne** [comment]
Just did another experiment:
From `readelf -d` on the `linux-musl-arm` binary:
```
 0x00000001 (NEEDED)                     Shared library: [libpthread.so.0]
 0x00000001 (NEEDED)                     Shared library: [libdl.so.2]
 0x00000001 (NEEDED)                     Shared library: [libstdc++.so.6]
 0x00000001 (NEEDED)                     Shared library: [libm.so.6]
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so.6]
 0x00000001 (NEEDED)                     Shared library: [ld-linux-armhf.so.3]
```

And on the `linux-musl-arm64` binary:
```
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.musl-aarch64.so.1]
```

So.. looks like the `musl-arm` RID is in fact compiling with glibc.

**Comment by janvorli** [comment]
We don't publish packages for linux-musl-arm yet, only linux-musl-arm64. So on arm, we seem to be falling back to linux-arm which is glibc.

**Comment by hamish-milne** [comment]
Ah I see - makes sense I suppose!

Anything I can do to ameliorate this issue? :) I'm using ghost's instructions to compile coreclr but perhaps things have moved on since then?

**Comment by ericstj** [comment]
Cc @akoeplinger needed to add arm64 to RID graph.  Perhaps we should add arm as well?  I think if you do that and build for arm it should work.

**Comment by danmosemsft** [comment]
@richlander are you tracking this issue? Do you see work here that is must have to ship a successful 5.0?

**Comment by danmosemsft** [comment]
@dleeapho do you have context on why we don't have Alpine ARM images? (32 bit)

cc @richlander

**Comment by janvorli** [comment]
After chatting with @richlander  last week, I have started to look into enabling official Alpine ARM builds. I have changes in the runtime, dotnet-buildtools-prereqs-docker and sdk repos prepared locally, I am currently waiting for a fix that I've made to arcade few days ago to propagate so that I can continue working on it. It seems it should be quite straightforward.

**Comment by danmosemsft** [comment]
Oh, that's good news. Thanks @janvorli !

**Comment by dleeapho** [comment]
@MichaelSimons @mthalman heads up on adding Alpine ARM32 dotnet-docker images once @janvorli is done.

**Comment by richlander** [comment]
Yes, @janvorli talked to me. This is not required for 5.0. It's a nice-to-have based on the hope of a low cost profile for the work. We also don't expect many users of this. 

In my mind, the real reason to do this isn't to enable support for Alpine ARM32 but that we'd now have Alpine support for all supported container configurations. This would enable us to create manifest tags for Alpine images, which would then make it easier for us to provide better UX for Alpine. So, in doing this work for ARM32, we can improve the experience for x64 and ARM64 (similar to what exists for Debian already).

**Comment by joperezr** [comment]
Pushing the milestone to 6.0 given @richlander comment above. We might still get it in for 5.0 but won't be tracked as a must-have.

**Comment by JensNordenbro** [comment]
I just would like to add as the creator of this item that there is much more to this than just symmetry amongst your rids. This is also about embedded support and the fact that all docker images for serious production use tend to gravitate towards Alpine. Basically this will make. Net core attractive on a new market!

**Comment by janvorli** [comment]
Update: I've merged in all necessary changes to the runtime repo, so nuget packages should be available soon. 
The next step is to enable producing runtime / sdk packages.

**Comment by mthalman** [comment]
@janvorli - Do you have an estimate of when this will be ready?  Will it be in RC2?

**Comment by janvorli** [comment]
@mthalman not in RC2, as the window got closed. I will try to request accepting port of the change to the GA though.

**Comment by JensNordenbro** [comment]
Hi @janvorli ! 

Thanks for the progress on this item! (best item ever for C# on Embedded Linux development  :) )

With the completion of #42963 **what is the state of this item**? 

on net sdk 5.0.101: 
'dotnet publish -r linux-musl-arm' seems to work (I have yet to validate the result on a linux system) 
..for both console and mvc projects. 

Looking at dotnet/installer Github repo I see no version tag, is it because things are still missing or just some other quirk ? 

![image](https://user-images.githubusercontent.com/9577365/105606551-5b1e6f00-5d9a-11eb-9fcf-ef399c3563d9.png)

**Comment by janvorli** [comment]
@JensNordenbro the missing version badges are just a quirk. 
Regarding the SDK, the 5.0.10x SDKs still have the linux-musl-arm broken and publish generates linux-arm files instead. The first SDK that works correctly is the 5.0.2xx preview.

**Comment by JensNordenbro** [comment]
Ok, good to know! 
Do I need to create a separate issue for enabling alpine containers or will that sort itself out in your processes?

**Comment by janvorli** [comment]
They were already created few days ago: https://github.com/dotnet/dotnet-docker/pull/2443

**Comment by JensNordenbro** [comment]
nice!

what is the condition to close this item? It seems done but not released to me!

**Comment by JensNordenbro** [comment]
Still not part of https://hub.docker.com/_/microsoft-dotnet-runtime/

**Comment by mthalman** [comment]
> Still not part of https://hub.docker.com/_/microsoft-dotnet-runtime/

@JensNordenbro - This is blocked by https://github.com/dotnet/runtime/issues/47423.

**Comment by ccifuentesfer** [comment]
Hello, 
We need to move to arm32 some of our IoT code, currently seems that the only option is going back to C :/
Is there any update on this issue? And also, is there any preview version we can use to test?

Best regards,

**Comment by mthalman** [comment]
@ccifuentesfer - Which version of .NET are you wanting? It's available in the latest servicing release of 5.0.  

@rbhanda - Can we get download links for Arm Alpine added to the [release notes](https://github.com/dotnet/core/blob/main/release-notes/5.0/5.0.7/5.0.7.md#downloads) going forward?

@janvorli - Can this issue be closed?

**Comment by janvorli** [comment]
Yes, it can be closed. There are no known remaining issues.

**Comment by ccifuentesfer** [comment]
Hello,
I have tried with versions:
5.0.301
6.0.100-preview.5.21302.13

with runtime linux-musl-arm and self-contained, the executable uses ld-linux-armhf.so.3 libc library.

Thanks for your reply!

**Comment by janvorli** [comment]
I've just tried with 5.0.301 SDK:
```sh
mkdir test
cd test
dotnet new console
dotnet publish -r linux-musl-arm --self-contained
```
And then:
```
objdump -x bin/Debug/net5.0/linux-musl-arm/publish/test | grep NEEDED
  NEEDED               libstdc++.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.musl-armv7.so.1
```
or
```
strings bin/Debug/net5.0/linux-musl-arm/publish/test | grep musl
/lib/ld-musl-armhf.so.1
libc.musl-armv7.so.1
```

**Comment by ccifuentesfer** [comment]
Thanks @janvorli it's working now!

The project had the target framework set to netcoreapp3.1, as soon as I changed it to net5.0 has been solved.

**Comment by janvorli** [comment]
@ccifuentesfer great, thank you for confirming that!
